Implementation of Server Reflection added for GRPC Server (with minor doc change)

PythonHacker24 opened the pull request:
I found this line in the docs about implementing server reflection in the propeller. 

![Screenshot 2024-12-22 at 22 58 20](https://github.com/user-attachments/assets/bcce210a-b304-42a2-92b3-8177b61fb8c0)

So I implemented a basic server reflection into the `newGrpcServer()` function in the file `/internal/grpcserver/server.go` and tested it on my machine (in Mac M2 environment). 

![Screenshot 2024-12-22 at 22 58 50](https://github.com/user-attachments/assets/6be70874-29a6-4aad-bd05-33dbb8dff458)

But I am concerned about whether I should add **access control**. Like in the production environment, access must be restricted to the reflection service and only allow authorized users/environments to access it. I am thinking of implementing a **middleware** that would check for an API Key to access the server reflection. Another way is to implement **IP restriction** (which I think is best for static IPs in internal environments). Or should I move on to implement a **config file** where any of these two methods can be selected. 

Let me know how I should move forward if the current change is acceptable. 

P.S. I made a small change in the docs in the context of clarity. 

Thanks.

Abhishekvrshny commented:
Hi @PythonHacker24 , thanks for the PR. 

I would prefer a config option like `EnableReflection` under `grpc` section in the [config file](https://github.com/CRED-CLUB/propeller/blob/main/config/propeller.toml#L25) as it's a more generic way to implement the same.

PythonHacker24 commented:
Welcome @Abhishekvrshny 

I have updated the code for the param named `EnableReflection` to toggle server reflection. I tested the Mac M2 environment code on `go 1.23`, which works fine. 

![Screenshot 2024-12-23 at 10 13 17](https://github.com/user-attachments/assets/37dd9a9a-6bfb-432a-ac96-4b82ab682585)

When I updated the code, the build wasn't working due to a cyclic dependency issue when I was running the code locally, so I had to comment out `github.com/cred-club/propeller/internal/config`, which worked fine, and I tested it. 

![Screenshot 2024-12-23 at 09 59 47](https://github.com/user-attachments/assets/1f34e524-db48-44a9-a536-8d2281120eb0)

Let me know how I can proceed with this.

Thanks.

PythonHacker24 commented:
I will update the docs later for `EnableReflection` due to merge conflicts occurring. I didn't sync my fork after yesterday's updates. Sorry for that.

Abhishekvrshny commented on docs/configurations.md:
```diff
@@ -8,30 +8,31 @@ nav_order: 7
 
 Configuration reference for `propeller.toml`.
 
-| Config                             | Values          | Description                                                                                                                                |
```
what is the change in this file?

Abhishekvrshny replied:
Please add documentation for this config here

PythonHacker24 replied:
I was adding the new option from `propeller.toml` to the docs. But since the docs were updated the day before yesterday and I didn't sync my fork, it caused a merge conflict. So I had to remove the changes to solve it.

PythonHacker24 replied:
Now I have synced the fork, got the code up to date, and made the changes in the docs.

Abhishekvrshny commented on internal/grpcserver/config.go:
```diff
@@ -5,4 +5,5 @@ type Config struct {
 	Address                  string
 	PingIntervalInSec        int
 	PingResponseTimeoutInSec int
+    EnableReflection         bool
```
please run `make goimports` to format the code

Abhishekvrshny replied:
@PythonHacker24 let me know once done

PythonHacker24 replied:
Done running `make go imports` and pushed the changes

Abhishekvrshny commented on go.sum:
```diff
@@ -67,6 +67,8 @@ github.com/census-instrumentation/opencensus-proto v0.2.1/go.mod h1:f6KPmirojxKA
 github.com/cespare/xxhash/v2 v2.1.1/go.mod h1:VGX0DQ3Q6kWi7AoAeZDth3/j3BFtOZR5XLFGgcrjCOs=
 github.com/cespare/xxhash/v2 v2.2.0 h1:DC2CZ1Ep5Y4k3ZQ899DldepgrayRUGE6BBZ/cd9Cj44=
 github.com/cespare/xxhash/v2 v2.2.0/go.mod h1:VGX0DQ3Q6kWi7AoAeZDth3/j3BFtOZR5XLFGgcrjCOs=
+github.com/cespare/xxhash/v2 v2.3.0 h1:UL815xU9SqsFlibzuggzjXhog7bL6oX9BbNZnL2UFvs=
```
you can also run `go mod tidy`

PythonHacker24 replied:
Done with `go mod tidy`

Abhishekvrshny commented on internal/grpcserver/server.go:
```diff
@@ -5,9 +5,11 @@ import (
 	"net"
 	"time"
 
+	// "github.com/CRED-CLUB/propeller/internal/config"
```
this can be removed

PythonHacker24 replied:
Done

Abhishekvrshny commented on config/propeller.toml:
```diff
@@ -23,6 +23,7 @@ EnableProfilingHandlers = false
     Address = "0.0.0.0:5011"
     PingIntervalInSec = 20
     PingResponseTimeoutInSec = 10
+    EnableReflection = true 
```
default value can be false

PythonHacker24 replied:
Done

Abhishekvrshny commented on docs/configurations.md:
```diff
@@ -28,6 +28,7 @@ Configuration reference for `propeller.toml`.
 | grpc.Address                       | string          | gRPC server port to bind.                                                                                                                   |
 | grpc.PingIntervalInSec             | integer         | gRPC keepalive configuration. [Reference](https://grpc.io/docs/guides/keepalive/)                                                           |
 | grpc.PingResponseTimeoutInSec      | integer         | gRPC keepalive configuration. [Reference](https://grpc.io/docs/guides/keepalive/)                                                           |
+| grpc.EnableReflection              | bool            | If server refection should be enabled for gRPC server                                                                                       |
```
while `bool` type is correct here, can we keep it consistent with other rows as `true/false`

PythonHacker24 replied:
Done

Abhishekvrshny commented:
Approved. But if you can also update the documentation [here](https://github.com/CRED-CLUB/propeller/blob/main/docs/testing.md?plain=1#L9) and mention to enable reflection through the config parameter  for testing, it would help

PythonHacker24 commented:
Done with updates the docs for te`testing.md` about server reflection

Abhishekvrshny commented:
Thanks for the contribution @PythonHacker24 !

PythonHacker24 commented:
@Abhishekvrshny Welcome 🙌
